dm/dm/master/scheduler/scheduler.go

```diff
@@ -2094,33 +2049,37 @@ func (s *Scheduler) deleteWorker(name string) {
 	metrics.RemoveWorkerState(w.baseInfo.Name)
 }
 
-// updateStatusForBound updates the in-memory status for bound, including:
+// updateStatusToBound updates the in-memory status for bound, including:
 // - update the stage of worker to `Bound`.
 // - record the bound relationship and last bound relationship in the scheduler.
+// - remove the unbound relationship in the scheduler.
 // this func is called after the bound relationship existed in etcd.
-// TODO: we could only let updateStatusForBound and updateStatusForUnbound to update s.unbounds/bounds/lastBound.
-func (s *Scheduler) updateStatusForBound(w *Worker, b ha.SourceBound) error {
+func (s *Scheduler) updateStatusToBound(w *Worker, b ha.SourceBound) error {
 	if err := w.ToBound(b); err != nil {
 		return err
 	}
 	s.bounds[b.Source] = w
 	s.lastBound[b.Worker] = b
+	delete(s.unbounds, b.Source)
 	return nil
 }
 
-// updateStatusForUnbound updates the in-memory status for unbound, including:
+// updateStatusToUnbound updates the in-memory status for unbound, including:
 // - update the stage of worker to `Free` or `Relay`.
 // - remove the bound relationship in the scheduler.
+// - record the unbound relationship in the scheduler.
 // this func is called after the bound relationship removed from etcd.
-func (s *Scheduler) updateStatusForUnbound(source string) {
+func (s *Scheduler) updateStatusToUnbound(source string) {
 	w, ok := s.bounds[source]
 	if !ok {
+		s.unbounds[source] = struct{}{}
```

Comment:
We can move this line before this if since all branches need to do this.

Reply:
https://github.com/pingcap/ticdc/pull/3257/commits/571f48dd2ceb04924b00bf37be3b48c21b2ba02e

PTAL